pkg/solana/logpoller/worker.go

```diff
@@ -0,0 +1,372 @@
+		Name:  "WorkerGroup",
+		Start: g.start,
+		Close: g.close,
+	}.NewServiceEngine(lggr)
+
+	for idx := range workers {
+		g.workers <- &worker{
+			Name:  fmt.Sprintf("worker-%d", idx+1),
+			Queue: g.workers,
+			Retry: g.chRetry,
+			Lggr:  g.lggr,
+		}
+	}
+
+	return g
+}
+
+var _ services.Service = &WorkerGroup{}
+
+func (g *WorkerGroup) start(ctx context.Context) error {
+	g.engine.Go(g.runQueuing)
+	g.engine.Go(g.runProcessing)
+	g.engine.Go(g.runRetryQueue)
+	g.engine.Go(g.runRetries)
+
+	return nil
+}
+
+func (g *WorkerGroup) close() error {
+	if !g.queueClosed.Load() {
+		g.queueClosed.Store(true)
+		close(g.chStopInputs)
+	}
+
+	return nil
+}
+
+// Do adds a new work item onto the work queue. This function blocks until
+// the work queue clears up or the context is cancelled. This allows a max wait
+// time for the queue to open. Or a context can wrap a collection of jobs that
+// need to be run and when the context cancels, the jobs don't get added to the
+// queue.
+func (g *WorkerGroup) Do(ctx context.Context, job Job) error {
+	if ctx.Err() != nil {
+		return fmt.Errorf("%w; work not added to queue", ErrContextCancelled)
+	}
+
+	if g.queueClosed.Load() {
+		return fmt.Errorf("%w; work not added to queue", ErrProcessStopped)
+	}
+
+	select {
+	case g.input <- job:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("%w; work not added to queue", ErrContextCancelled)
+	case <-g.chStopInputs:
+		return fmt.Errorf("%w; work not added to queue", ErrProcessStopped)
+	}
+}
+
+func (g *WorkerGroup) runQueuing(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-g.input:
+			g.queue.Add(item)
+
+			// notify that new work item came in
+			// drop if notification channel is full
+			select {
+			case g.chInputNotify <- struct{}{}:
+			default:
+			}
+		}
+	}
+}
+
+func (g *WorkerGroup) runProcessing(ctx context.Context) {
+Loop:
+	for {
+		select {
+		// watch notification channel and begin processing queue
+		// when notification occurs
+		case <-g.chInputNotify:
+			g.processQueue(ctx)
+		case <-ctx.Done():
+			break Loop
+		}
+	}
+
+	// run the job queue one more time just in case some
+	// new work items snuck in
+	g.processQueue(ctx)
+}
+
+func (g *WorkerGroup) runRetryQueue(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-g.chRetry:
+			var retry retryableJob
+
+			switch typedJob := job.(type) {
+			case retryableJob:
+				retry = typedJob
+				retry.count++
+
+				if retry.count > g.maxRetryCount {
+					g.lggr.Infof("job %s dropped after max retries", job)
+
+					continue
+				}
+
+				wait := calculateExponentialBackoff(retry.count)
+				g.lggr.Infof("retrying job in %dms", wait/time.Millisecond)
+
+				retry.when = time.Now().Add(wait)
+			default:
+				wait := calculateExponentialBackoff(0)
+
+				g.lggr.Infof("retrying job %s in %dms", job, wait/time.Millisecond)
+
+				retry = retryableJob{
+					name: createRandomString(12),
```

Comment:
I understand the uint64 counter which could save some memory footprint, but I don't understand using a slice to save allocations. The retry still needs the job and the time in which to retry. This could possibly be in a `map[time.Time]Job` where the key is the time when the job should be retried. At that point the random name isn't relevant.

Reply:
I was thinking that we can do something like
```go
var retries []Job
for i := 0; i < len(g.retrySlice); {
  retry := g.retrySlice[i]
  if time.Now().After(retry.when) {
    retries = append(retries, retry)
     g.retrySlice[i] =  g.retrySlice[len(g.retrySlice)-1] 
     g.retrySlice  = g.retrySlice[:len(g.retrySlice)-1] 
   } else {
     i++
   }
}
```

But now realized that it could cause dead lock.

Reply:
@EasterTheBunny, what are your thoughts? Should not we at least drop string generation?